python.md

```diff
@@ -313,6 +314,51 @@ Here's a few tips.
 
   Note the `safe` tag which allow HTML to be included in the message.
 
+### <a name="model-forms">Avoid model forms</a>
+
+[Django's model-forms](https://docs.djangoproject.com/en/2.1/topics/forms/modelforms/) are a useful crutch for rapidly building web projects.
+However for a late-stage Django project (like ours), they are best avoided apart from the
+most degenerate, simple scenarios.
+
+Why? Because they conflate validation and persistence logic which, over time,
+leads to hard-to-maintain code. As soon as you need to add more sophisticated
+actions (than a simple DB write) to a successful form submission, model-forms
+are the wrong choice.  Once you start overriding `.save()`, you're on the path
+to maintenance hell.  Future maintainers will thank you if you ensure forms are
+only responsible for validating dictionaries of data, nothing more,
+single-responsibility principle and all that.
+
+Instead, use plain subclasses of `form.Form` and Django's `fields_for_model`
+function to extract the form fields you need. Handle the `form_valid` scenario
```

Comment:
An alternative pattern is to call the domain from form methods instead of directly from the view (see octoenergy.interfaces.supportsite.billing.views.EditMeterPointTaxes as an example, which leverages the fact that UpdateView calls form.save() automatically). I have a slight preference for doing this as it puts the calling of the domain close to the validation of data (and keeps views for request/response logic). I also like views to be as small as possible.

Do you think it would be better to establish a convention one way or the other, or keep it flexible (as either pattern is fine). I don't feel strongly, but either way, maybe it's worth making it explicit here?

Reply:
I personally don't like having `save` methods on forms as it's a SRP violation (and can lead to hard-to-maintain code over time). I find things easier to reason about if forms are only responsible for providing clean, valid data. 

A corollary of this is there's not that much point using generic views, it's better to stick to `FormView`, `TemplateView` and `RedirectView`s even though it's a bit more verbose (and more work upfront) than using the ORM-based views/forms. It's a better path for writing maintainable code in the long term.

Reply:
Yep makes sense. It might be worth adding this to the style guide at some point.